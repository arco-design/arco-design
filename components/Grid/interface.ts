import { CSSProperties, HTMLAttributes, ReactNode } from 'react';

export type GridResponsiveBreakpoint = 'xxxl' | 'xxl' | 'xl' | 'lg' | 'md' | 'sm' | 'xs';

export type GridRowGutter = number | Partial<Record<GridResponsiveBreakpoint, number>>;

/**
 * @title Row
 */
export interface RowProps extends Omit<HTMLAttributes<HTMLDivElement>, 'className'> {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh
   * 栅格间隔，单位是`px` 栅格间隔。可传入响应式对象写法 { xs: 4, sm: 6, md: 12}，传入数组 [ 水平间距， 垂直间距 ] 来设置两个方向。
   * @en
   * Spacing between grids, could be a number or a object like { xs: 8, sm: 16, md: 24}.
   * Or you can use array to make horizontal and vertical spacing work at the same time [horizontal, vertical]
   * @version vertical gutter in 2.5.0
   * @defaultValue 0
   */
  gutter?: GridRowGutter | Array<GridRowGutter>;
  /**
   * @zh 开启这个选项 `<Row>` 和 `<Col>` 都会被当作 div 而不会附带任何 Grid 相关的类和样式
   * @en If true, `<Row>` and `<Col>` will be treated as a div without any Grid related classes and styles
   */
  div?: boolean;
  /**
   * @zh 竖直对齐方式 ( `align-items` )
   * @en Vertical alignment, same as css `align-items`
   * @defaultValue start
   */
  align?: 'start' | 'center' | 'end' | 'stretch';
  /**
   * @zh 水平对齐方式 (`justify-content`)
   * @en Horizontal alignment, same as css `justify-content`
   * @defaultValue start
   */
  justify?: 'start' | 'center' | 'end' | 'space-around' | 'space-between';
}

/**
 * @title Col
 */
export interface ColProps extends Omit<HTMLAttributes<HTMLDivElement>, 'className' | 'ref'> {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 栅格占位格数
   * @en Raster number of cells to occupy
   * @defaultValue 24
   */
  span?: number;
  /**
   * @zh 栅格左侧的间隔格数，间隔内不可以有栅格
   * @en The number of cells to offset Col from the left
   */
  offset?: number;
  /**
   * @zh 对元素进行排序
   * @en Raster order
   */
  order?: number;
  /**
   * @zh 对元素进行排序
   * @en Raster order
   * @version 2.20.0
   */
  push?: number;
  /**
   * @zh 对元素进行排序
   * @en Raster order
   * @version 2.20.0
   */
  pull?: number;
  /**
   * @zh < 576px 响应式栅格
   * @en `screen < 576px`
   */
  xs?: number | { [key: string]: any };
  /**
   * @zh >= 576px 响应式栅格
   * @en `screen >= 576px`
   */
  sm?: number | { [key: string]: any };
  /**
   * @zh >= 768px 响应式栅格
   * @en `screen >= 768px`
   */
  md?: number | { [key: string]: any };
  /**
   * @zh >= 992px 响应式栅格
   * @en `screen >= 992px`
   */
  lg?: number | { [key: string]: any };
  /**
   * @zh >= 1200px 响应式栅格
   * @en `screen >= 1200px`
   */
  xl?: number | { [key: string]: any };
  /**
   * @zh >= 1600px 响应式栅格
   * @en `screen >= 1600px`
   */
  xxl?: number | { [key: string]: any };
  /**
   * @zh >= 2000px 响应式栅格
   * @en `screen >= 2000px`
   * @version 2.40.0
   */
  xxxl?: number | { [key: string]: any };
  /**
   * @zh 设置 flex 布局属性
   * @en Set flex layout properties
   * @version 2.26.0
   */
  flex?: FlexType;
}

/**
 * @title Grid
 */
export interface GridProps extends Omit<HTMLAttributes<HTMLDivElement>, 'className'> {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 每一行展示的列数
   * @en Number of columns displayed in each row
   * @defaultValue 24
   */
  cols?: number | ResponsiveValue;
  /**
   * @zh 行与行之间的间距
   * @en The space in row-to-row
   * @defaultValue 0
   */
  rowGap?: number | ResponsiveValue;
  /**
   * @zh 列与列之间的间距
   * @en The space in column-to-column
   * @defaultValue 0
   */
  colGap?: number | ResponsiveValue;
  /**
   * @zh 是否折叠
   * @en Whether to collapsed
   * @defaultValue false
   */
  collapsed?: boolean;
  /**
   * @zh 折叠时显示的行数
   * @en Number of rows displayed when collapsed
   * @defaultValue 1
   */
  collapsedRows?: number;
}

/**
 * @title GridItem
 */
export interface GridItemProps extends Omit<HTMLAttributes<HTMLDivElement>, 'className'> {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 跨越的格数
   * @en Number of grids spanned
   * @defaultValue 1
   */
  span?: number | ResponsiveValue;
  /**
   * @zh 左侧的间隔格数
   * @en Number of grids on the left
   * @defaultValue 0
   */
  offset?: number | ResponsiveValue;
  /**
   * @zh 是否是后缀元素
   * @en Is it a suffix element
   * @defaultValue false
   */
  suffix?: boolean;
  __index__?: number;
  children?: ReactNode | ((info: { overflow: boolean }) => ReactNode);
}

export interface GridItemData extends GridItemProps {
  span: number;
  offset: number;
}

/**
 * @title ResponsiveValue
 */
export interface ResponsiveValue {
  /**
   * @zh >= 2000px 响应式栅格
   * @en `screen >= 2000px`
   */
  xxxl?: number;
  /**
   * @zh >= 1600px 响应式配置
   * @en >= 1600px responsive configuration
   */
  xxl?: number;
  /**
   * @zh >= 1200px 响应式配置
   * @en >= 1200px responsive configuration
   */
  xl?: number;
  /**
   * @zh >= 992px 响应式配置
   * @en >= 992px responsive configuration
   */
  lg?: number;
  /**
   * @zh >= 768px 响应式配置
   * @en >= 768px responsive configuration
   */
  md?: number;
  /**
   * @zh >= 576px 响应式配置
   * @en >= 576px responsive configuration
   */
  sm?: number;
  /**
   * @zh < 576px 响应式配置
   * @en < 576px responsive configuration
   */
  xs?: number;
}

export type FlexType = string | number | 'auto' | 'none';
