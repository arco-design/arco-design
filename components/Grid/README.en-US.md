`````
Component / Layout

# Grid

Grid can effectively ensure the consistency and logic of the page, strengthen teamwork and unity.
`````

%%Content%%

## API

### Row

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|style|Additional style|`CSSProperties`|`-`|-|
|className|Additional css class|`string \| string[]`|`-`|-|
|gutter|Spacing between grids, could be a number or a object like { xs: 8, sm: 16, md: 24}.Or you can use array to make horizontal and vertical spacing work at the same time [horizontal, vertical]|`GridRowGutter \| Array<GridRowGutter>`|`0`|vertical gutter in 2.5.0|
|div|If true, `<Row>` and `<Col>` will be treated as a div without any Grid related classes and styles|`boolean`|`-`|-|
|align|Vertical alignment, same as css `align-items`|`'start' \| 'center' \| 'end' \| 'stretch'`|`start`|-|
|justify|Horizontal alignment, same as css `justify-content`|`'start' \| 'center' \| 'end' \| 'space-around' \| 'space-between'`|`start`|-|

### Col

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|style|Additional style|`CSSProperties`|`-`|-|
|className|Additional css class|`string \| string[]`|`-`|-|
|span|Raster number of cells to occupy|`number`|`24`|-|
|offset|The number of cells to offset Col from the left|`number`|`-`|-|
|order|Raster order|`number`|`-`|-|
|push|Raster order|`number`|`-`|2.20.0|
|pull|Raster order|`number`|`-`|2.20.0|
|xs|`screen < 576px`|`number \| { [key: string]: any }`|`-`|-|
|sm|`screen >= 576px`|`number \| { [key: string]: any }`|`-`|-|
|md|`screen >= 768px`|`number \| { [key: string]: any }`|`-`|-|
|lg|`screen >= 992px`|`number \| { [key: string]: any }`|`-`|-|
|xl|`screen >= 1200px`|`number \| { [key: string]: any }`|`-`|-|
|xxl|`screen >= 1600px`|`number \| { [key: string]: any }`|`-`|-|
|flex|Set flex layout properties|`FlexType`|`-`|2.26.0|

```ts
type FlexType = number | string | 'auto' | 'none';
```
