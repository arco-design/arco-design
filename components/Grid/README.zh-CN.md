`````
组件 / 布局

# 栅格 Grid

栅格可以有效的保证页面的一致性、逻辑性、加强团队协作和统一。
`````

%%Content%%

## API

### Row

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|style|节点样式|`CSSProperties`|`-`|-|
|className|节点类名|`string \| string[]`|`-`|-|
|gutter|栅格间隔，单位是`px` 栅格间隔。可传入响应式对象写法 { xs: 4, sm: 6, md: 12}，传入数组 [ 水平间距， 垂直间距 ] 来设置两个方向。|`GridRowGutter \| Array<GridRowGutter>`|`0`|vertical gutter in 2.5.0|
|div|开启这个选项 `<Row>` 和 `<Col>` 都会被当作 div 而不会附带任何 Grid 相关的类和样式|`boolean`|`-`|-|
|align|竖直对齐方式 ( `align-items` )|`'start' \| 'center' \| 'end' \| 'stretch'`|`start`|-|
|justify|水平对齐方式 (`justify-content`)|`'start' \| 'center' \| 'end' \| 'space-around' \| 'space-between'`|`start`|-|

### Col

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|style|节点样式|`CSSProperties`|`-`|-|
|className|节点类名|`string \| string[]`|`-`|-|
|span|栅格占位格数|`number`|`24`|-|
|offset|栅格左侧的间隔格数，间隔内不可以有栅格|`number`|`-`|-|
|order|对元素进行排序|`number`|`-`|-|
|push|对元素进行排序|`number`|`-`|2.20.0|
|pull|对元素进行排序|`number`|`-`|2.20.0|
|xs|< 576px 响应式栅格|`number \| { [key: string]: any }`|`-`|-|
|sm|>= 576px 响应式栅格|`number \| { [key: string]: any }`|`-`|-|
|md|>= 768px 响应式栅格|`number \| { [key: string]: any }`|`-`|-|
|lg|>= 992px 响应式栅格|`number \| { [key: string]: any }`|`-`|-|
|xl|>= 1200px 响应式栅格|`number \| { [key: string]: any }`|`-`|-|
|xxl|>= 1600px 响应式栅格|`number \| { [key: string]: any }`|`-`|-|
|flex|设置 flex 布局属性|`FlexType`|`-`|2.26.0|

```ts
type FlexType = number | string | 'auto' | 'none';
```
