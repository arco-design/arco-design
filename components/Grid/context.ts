import { createContext } from 'react';
import { GridItemData } from './interface';

type RowContextType = {
  gutter?: [number, number];
  div?: boolean;
};

export const RowContext = createContext<RowContextType>({});

type GridContextType = {
  overflow?: boolean;
  collapsed?: boolean;
  displayIndexList?: number[];
  cols?: number;
  colGap?: number;
};

export const GridContext = createContext<GridContextType>({});

type GridDataCollectorType = Readonly<{
  collectItemData?: (index: number, itemData: GridItemData) => void;
  removeItemData?: (index: number) => void;
}>;

export const GridDataCollectorContext = createContext<GridDataCollectorType>({});
