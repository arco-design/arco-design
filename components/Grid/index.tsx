import Col from './col';
import Row from './row';

export default {
  Col,
  Row,
};

export type { RowProps, ColProps } from './interface';
