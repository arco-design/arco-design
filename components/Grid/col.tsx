import React, { useContext, useMemo, forwardRef } from 'react';
import cs from '../_util/classNames';
import { isNumber, isObject } from '../_util/is';
import { ConfigContext } from '../ConfigProvider';
import { ColProps, FlexType } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';
import { RowContext } from './context';

const defaultProps: ColProps = {
  span: 24,
};

function getFlexString(flex: FlexType) {
  if (typeof flex === 'string' && /\d+[px|%|em|rem|]{1}/.test(flex)) {
    return `0 0 ${flex}`;
  }
  return flex;
}

function Col(baseProps: ColProps, ref) {
  const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<ColProps>(baseProps, defaultProps, componentConfig?.['Grid.Col']);
  const { gutter, div } = useContext(RowContext);

  const {
    className,
    style,
    children,
    span,
    offset,
    order,
    pull,
    push,
    xs,
    sm,
    md,
    lg,
    xl,
    xxl,
    xxxl,
    flex,
    ...rest
  } = props;

  function adaptationGrid(prefixCls: string, mergeClassName: { [key: string]: any }) {
    const screenList = { xs, sm, md, lg, xl, xxl, xxxl };
    Object.keys(screenList).forEach((screen) => {
      const screenValue = screenList[screen];
      if (isNumber(screenValue)) {
        if (screenValue >= 0) {
          mergeClassName[`${prefixCls}-${screen}-${screenValue}`] = true;
        }
      } else if (isObject(screenValue)) {
        mergeClassName[`${prefixCls}-${screen}-${screenValue.span}`] = screenValue.span;
        mergeClassName[`${prefixCls}-${screen}-offset-${screenValue.offset}`] = screenValue.offset;
        mergeClassName[`${prefixCls}-${screen}-order-${screenValue.order}`] = screenValue.order;
        mergeClassName[`${prefixCls}-${screen}-pull-${screenValue.pull}`] = screenValue.pull;
        mergeClassName[`${prefixCls}-${screen}-push-${screenValue.push}`] = screenValue.push;
      }
    });
    return mergeClassName;
  }

  const prefixCls = getPrefixCls('col');
  let mergeClassName = {
    [`${prefixCls}`]: !div,
    [`${prefixCls}-order-${order}`]: order,
    [`${prefixCls}-${span}`]: !div && !xs && !sm && !md && !lg && !xl && !xxl && !xxxl,
    [`${prefixCls}-offset-${offset}`]: offset,
    [`${prefixCls}-pull-${pull}`]: pull,
    [`${prefixCls}-push-${push}`]: push,
    [`${prefixCls}-rtl`]: rtl,
  };
  mergeClassName = adaptationGrid(prefixCls, mergeClassName);
  const classNames = cs(flex ? prefixCls : mergeClassName, className);

  const paddingStyle: {
    paddingLeft?: number;
    paddingRight?: number;
    paddingTop?: number;
    paddingBottom?: number;
  } = {};
  if (Array.isArray(gutter) && !div) {
    const paddingHorizontal = (gutter[0] && gutter[0] / 2) || 0;
    const paddingVertical = (gutter[1] && gutter[1] / 2) || 0;
    if (paddingHorizontal) {
      paddingStyle.paddingLeft = paddingHorizontal;
      paddingStyle.paddingRight = paddingHorizontal;
    }
    if (paddingVertical) {
      paddingStyle.paddingTop = paddingVertical;
      paddingStyle.paddingBottom = paddingVertical;
    }
  }

  const flexStyle = useMemo(
    () => (getFlexString(flex) ? { flex: getFlexString(flex) } : {}),
    [flex]
  );

  return (
    <div
      ref={ref}
      {...rest}
      style={{
        ...style,
        ...paddingStyle,
        ...flexStyle,
      }}
      className={classNames}
    >
      {children}
    </div>
  );
}

const ColComponent = forwardRef(Col);

ColComponent.displayName = 'Col';

export default ColComponent;

export { ColProps };
