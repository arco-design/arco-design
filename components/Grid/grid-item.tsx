import React, { forwardRef, ReactNode, useContext, useEffect, useMemo } from 'react';
import { GridItemProps } from './interface';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import useMergeProps from '../_util/hooks/useMergeProps';
import { useResponsiveState } from './hooks/useResponsiveState';
import { GridContext, GridDataCollectorContext } from './context';
import { resolveItemData } from './utils';
import { isFunction, isString } from '../_util/is';

const defaultProps: GridItemProps = {
  suffix: false,
  offset: 0,
  span: 1,
};

function GridItem(baseProps: GridItemProps, ref) {
  const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<GridItemProps>(
    baseProps,
    defaultProps,
    componentConfig?.['Grid.GridItem']
  );

  const {
    children,
    className,
    style,
    offset: propOffset,
    span: propSpan,
    __index__: computedIndex,
  } = props;
  const gridContext = useContext(GridContext);

  const { collectItemData, removeItemData } = useContext(GridDataCollectorContext);

  const { colGap, cols, displayIndexList, overflow } = gridContext;

  const offset = useResponsiveState(propOffset, 0);
  const span = useResponsiveState(propSpan, 1);

  const prefixCls = getPrefixCls('grid-item');

  const visible = displayIndexList?.includes(computedIndex);

  const mergeClassName = {
    [`${prefixCls}`]: true,
    [`${prefixCls}-rtl`]: rtl,
  };

  const classNames = cs(mergeClassName, className);

  const itemData = useMemo(() => {
    return resolveItemData(gridContext.cols, {
      suffix: !!props.suffix,
      span,
      offset,
    });
  }, [gridContext.cols, props.suffix, span, offset]);

  useEffect(() => {
    collectItemData(computedIndex, itemData);

    return () => {
      removeItemData(computedIndex);
    };
  }, [computedIndex, itemData]);

  const offsetStyle = useMemo(() => {
    const { offset, span } = itemData;
    if (offset > 0) {
      const oneSpan = `(100% - ${colGap * (span - 1)}px) / ${span}`;
      return {
        marginLeft: `calc((${oneSpan} * ${offset}) + ${colGap * offset}px)`,
      };
    }
    return {};
  }, [itemData, colGap]);

  const columnStart = useMemo(() => {
    const { suffix, span } = itemData;
    if (suffix) {
      return `${cols - span + 1}`;
    }
    return `span ${span}`;
  }, [itemData, cols]);

  const visibleStyle = !visible || span === 0 ? { display: 'none' } : {};

  const gridItemStyle = {
    gridColumn: `${columnStart} / span ${span}`,
    ...offsetStyle,
    ...visibleStyle,
  };

  return (
    <div
      ref={ref}
      className={classNames}
      style={{
        ...gridItemStyle,
        ...style,
      }}
    >
      {isFunction(children)
        ? children({ overflow })
        : React.Children.map(children, (child: ReactNode) => {
            if (
              child &&
              gridContext.collapsed &&
              React.isValidElement(child) &&
              !isString(child.type)
            ) {
              // 排除原生 dom 标签，避免 overflow 属性透传到 dom 标签上
              return React.cloneElement(child, {
                overflow,
                ...child.props,
              });
            }
            return child;
          })}
    </div>
  );
}
const ForwardRefGridItem = forwardRef(GridItem);

const GridItemComponent = ForwardRefGridItem as typeof ForwardRefGridItem & {
  __ARCO_GRID_ITEM__: boolean;
};

GridItemComponent.displayName = 'GridItem';

GridItemComponent.__ARCO_GRID_ITEM__ = true;

export default GridItemComponent;

export { GridItemProps };
