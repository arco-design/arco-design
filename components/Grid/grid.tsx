import React, { forwardRef, ReactElement, useContext, useState } from 'react';
import get from 'lodash/get';
import { GridItemData, GridProps } from './interface';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import useMergeProps from '../_util/hooks/useMergeProps';
import { useResponsiveState } from './hooks/useResponsiveState';
import { GridContext, GridDataCollectorContext } from './context';
import { setItemVisible } from './utils';

const defaultProps: GridProps = {
  collapsed: false,
  collapsedRows: 1,
  cols: 24,
  colGap: 0,
  rowGap: 0,
};

function Grid(baseProps: GridProps, ref) {
  const [itemDataMap, setItemDataMap] = useState<Map<number, GridItemData>>(new Map());
  const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<GridProps>(baseProps, defaultProps, componentConfig?.Grid);
  const {
    children,
    className,
    style,
    cols: propCols,
    colGap: propColGap,
    rowGap: propRowGap,
    collapsed,
    collapsedRows,
  } = props;

  const cols = useResponsiveState(propCols, 24);
  const colGap = useResponsiveState(propColGap, 0);
  const rowGap = useResponsiveState(propRowGap, 0);

  const gridStyle = {
    gap: `${rowGap}px ${colGap}px`,
    gridTemplateColumns: `repeat(${cols}, minmax(0px, 1fr))`,
  };

  const prefixCls = getPrefixCls('grid');
  const mergeClassName = {
    [`${prefixCls}`]: true,
    [`${prefixCls}-rtl`]: rtl,
  };

  const classNames = cs(mergeClassName, className);

  const getItemDataList = () => {
    const list: GridItemData[] = [];
    for (const [index, itemData] of itemDataMap.entries()) {
      list[index] = itemData;
    }
    return list;
  };
  const itemDataList = getItemDataList();

  const displayInfo = setItemVisible({
    cols,
    collapsed,
    collapsedRows,
    itemDataList,
  });
  return (
    <div
      ref={ref}
      className={classNames}
      style={{
        ...gridStyle,
        ...style,
      }}
    >
      <GridDataCollectorContext.Provider
        value={{
          collectItemData(index, itemData) {
            itemDataMap.set(index, itemData);
            setItemDataMap(new Map([...itemDataMap]));
          },
          removeItemData(index) {
            itemDataMap.delete(index);
            setItemDataMap(new Map([...itemDataMap]));
          },
        }}
      >
        <GridContext.Provider
          value={{
            cols,
            colGap,
            collapsed,
            overflow: displayInfo.overflow,
            displayIndexList: displayInfo.displayIndexList,
          }}
        >
          {React.Children.map(children, (child: ReactElement, index) => {
            if (child) {
              const childProps = {
                __index__: index,
                ...child.props,
              };

              return React.cloneElement(child, childProps);
            }
            return null;
          }).filter((child: ReactElement) => get(child, 'type.__ARCO_GRID_ITEM__'))}
        </GridContext.Provider>
      </GridDataCollectorContext.Provider>
    </div>
  );
}

const GridComponent = forwardRef(Grid);

GridComponent.displayName = 'Grid';

export default GridComponent;

export { GridProps };
