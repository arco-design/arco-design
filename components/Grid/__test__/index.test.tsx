import React from 'react';
import { mount } from 'enzyme';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Grid from '..';

const Row = Grid.Row;
const Col = Grid.Col;

mountTest(Row);
mountTest(Col);
componentConfigTest(Grid.Row, 'Grid.Row');
componentConfigTest(Grid.Col, 'Grid.Col');

it('render text and gutter correctly', () => {
  const grid = mount(
    <Row gutter={12}>
      <Col span={12}>
        <div>Col-span-12 gutter-12</div>
      </Col>
      <Col span={12}>
        <div>Col-span-12 gutter-12</div>
      </Col>
    </Row>
  );

  grid.find('.arco-col').forEach((node) => {
    expect(node.prop('style')).toEqual({ paddingLeft: 6, paddingRight: 6 });
  });
});

it('gutter correct when row children is custom component', () => {
  function CustomCol(props) {
    const { children, ...rest } = props;
    return (
      <Col {...rest}>
        <div>{children}</div>
      </Col>
    );
  }

  const grid = mount(
    <Row gutter={12}>
      <CustomCol span={12}>
        <div>Col-span-12 gutter-12</div>
      </CustomCol>
      <CustomCol span={12}>
        <div>Col-span-12 gutter-12</div>
      </CustomCol>
    </Row>
  );

  grid.find('.arco-col').forEach((node) => {
    expect(node.prop('style')).toEqual({ paddingLeft: 6, paddingRight: 6 });
  });
});

it('flex', () => {
  const grid = mount(
    <Row>
      <Col flex="100px">100px</Col>
      <Col flex="auto">auto</Col>
      <Col flex="1 1 auto">1 1 auto</Col>
    </Row>
  );

  expect(grid.find('.arco-col').at(0).prop('style')).toEqual({ flex: '0 0 100px' });
  expect(grid.find('.arco-col').at(1).prop('style')).toEqual({ flex: 'auto' });
  expect(grid.find('.arco-col').at(2).prop('style')).toEqual({ flex: '1 1 auto' });
});
