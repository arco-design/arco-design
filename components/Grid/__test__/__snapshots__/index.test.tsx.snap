// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Grid.Col default 1`] = `
<div
  class="arco-col arco-col-24"
/>
`;

exports[`ConfigProvider componentConfig.Grid.Col set className globally 1`] = `
<div
  class="arco-col arco-col-24 global-default-class"
/>
`;

exports[`ConfigProvider componentConfig.Grid.Col set className globally and component set className itself. 1`] = `
<div
  class="arco-col arco-col-24 component-class"
/>
`;

exports[`ConfigProvider componentConfig.Grid.Col set data-* property 1`] = `
<div
  class="arco-col arco-col-24"
  data-test-name="Grid.Col"
/>
`;

exports[`ConfigProvider componentConfig.Grid.Row default 1`] = `
<div
  class="arco-row arco-row-align-start arco-row-justify-start"
/>
`;

exports[`ConfigProvider componentConfig.Grid.Row set className globally 1`] = `
<div
  class="arco-row arco-row-align-start arco-row-justify-start global-default-class"
/>
`;

exports[`ConfigProvider componentConfig.Grid.Row set className globally and component set className itself. 1`] = `
<div
  class="arco-row arco-row-align-start arco-row-justify-start component-class"
/>
`;

exports[`ConfigProvider componentConfig.Grid.Row set data-* property 1`] = `
<div
  class="arco-row arco-row-align-start arco-row-justify-start"
  data-test-name="Grid.Row"
/>
`;
